File: src/responses/embeds/nickname.rs
use chrono::Utc;

use poise::serenity_prelude as serenity;

use serenity::{
    colours::{
        branding::BLURPLE,
        css::{DANGER, POSITIVE},
    },
    *,
};

pub fn acknowledge_request() -> impl FnOnce(&mut CreateEmbed) -> &mut CreateEmbed {
    |embed| {
        embed
            .title("Nickname request")
            .colour(BLURPLE)
            .description(
                "Your nickname request has been received. \
                Please wait for approval by a mentor.",
            )
            .timestamp(Utc::now())
    }
}

pub fn request_approval<'a>(
    user_id: UserId,
    name: &'a str,
    current_nickname: &'a str,
    requested_nickname: &'a str,
) -> impl FnOnce(&mut CreateEmbed) -> &mut CreateEmbed + 'a {
    move |embed| {
        embed
            .title("Nickname request")
            .colour(BLURPLE)
            .description(format!("<@{}> has requested a new nickname.", user_id))
            .field("Current nickname", current_nickname, true)
            .field("Requested nickname", requested_nickname, true)
            .field("Verified name", name, false)
            .timestamp(Utc::now())
    }
}

pub fn approved(nickname: &str) -> impl FnOnce(&mut CreateEmbed) -> &mut CreateEmbed + '_ {
    move |embed| {
        embed
            .title("Nickname request")
            .colour(POSITIVE)
            .description(format!(
                "Your request to change your nickname to **{}** was **approved**.",
                nickname,
            ))
            .timestamp(Utc::now())
    }
}

pub fn denied(nickname: &str) -> impl FnOnce(&mut CreateEmbed) -> &mut CreateEmbed + '_ {
    move |embed| {
        embed
            .title("Nickname request")
            .colour(DANGER)
            .description(format!(
                "Your request to change your nickname to **{}** was **denied**. \
                Please note that you can message a mentor if you think this was a mistake.",
                nickname,
            ))
            .timestamp(Utc::now())
    }
}

File: norris/responses/embeds/nickname.py
from datetime import datetime

from discord import Colour, Embed, EmbedField


def acknowledge_request() -> Embed:
    return Embed(
        title="Nickname request",
        colour=Colour.blurple(),
        description="Your nickname request has been received. Please wait for "
                    "approval by a mentor.",
        timestamp=datetime.utcnow(),
    )


def request_approval(user_id: int,
                     name: str,
                     current_nickname: str,
                     requested_nickname: str) -> Embed:
    return Embed(
        title="Nickname request",
        colour=Colour.blurple(),
        description=f"<@{user_id}> has requested a new nickname.",
        fields=[
            EmbedField("Current nickname", current_nickname, inline=True),
            EmbedField("Requested nickname", requested_nickname, inline=True),
            EmbedField("Verified name", name),
        ],
        timestamp=datetime.utcnow(),
    )


def approved(nickname: str) -> Embed:
    return Embed(
        title="Nickname request",
        colour=Colour.green(),
        description=f"Your request to change your nickname to **{nickname}** was "
                    f"**approved**.",
        timestamp=datetime.utcnow(),
    )


def denied(nickname: str) -> Embed:
    return Embed(
        title="Nickname request",
        colour=Colour.red(),
        description=f"Your request to change your nickname to **{nickname}** was "
                    f"**denied**. Please note that you can message a mentor if you "
                    f"think this was a mistake.",
        timestamp=datetime.utcnow(),
    )
